### PR TITLE
Change pipeline for exporting to toffeeIR

### PR DESCRIPTION
**Pull request by zdevito**
Previously:
  PythonOp/CppOp Graph -> ToffeeIR, primspecs worked with protobufs
now:
  PythonOp/CppOp --ToToffeIR--> jit::Graph of in-memory ToffeIR -> protobufs of ToffeIR

This commit let's primspec functions work directly with JIT IR nodes,
which makes it possible to do a lot more stuff in those functions.

I can't build on the jit branch anymore, so contbuilds will likely fail, but I still would like to get this reviewed anticipating being able to make the build work again.

### CODE REVIEW

**Comment by zdevito** [comment]
I put the expect changes in a different commit because they are so large.

**Comment by ezyang** [comment]
batch_normalization.cpp doesn't seem to have been updated, which is now causing build failures for me (not caught by CI because CI is still running without Toffee.)

**Comment by ezyang** [comment]
I resolved the conflict but the build still fails due to batch norm and convolution not being ported to new format.

File: torch/autograd/_functions/basic_ops.py [hunk 1]
```diff
@@ -8,10 +8,10 @@
 class Add(InplaceFunction):
 
     @staticmethod
-    def primspec(a, b, inplace=False):
+    def primspec(g, a, b, inplace=False):
         if inplace:
             return None
-        return torch.toffee.op("Add", a, b)
+        return g.appendNode(g.create("Add", [a, b]))
```
Comment (apaszke) [inline torch/autograd/_functions/basic_ops.py:14]: How about adding `appendNew` equivalent to `appendNode(create(`? 😊

File: torch/autograd/_functions/basic_ops.py [hunk 2]
```diff
@@ -229,10 +229,10 @@ def backward(ctx, grad_output):
 class Negate(InplaceFunction):
 
     @staticmethod
-    def primspec(i, inplace=False):
+    def primspec(g, i, inplace=False):
         if inplace:
             return None
-        return torch.toffee.op("Scale", i, scale=float(-1))
+        return g.appendNode(g.create("Scale", [i]).f_("scale", -1))
```
Comment (apaszke) [inline torch/autograd/_functions/basic_ops.py:235]: It would be nice if we had a function that can do attribute setting automatically. kwarg lists are sooo much more readable than chains of `f_(...).i_(...).s_(...).i_(...)`

File: torch/autograd/_functions/tensor.py [hunk 3]
```diff
@@ -95,8 +96,11 @@ def backward(ctx, grad_output):
 class View(Function):
 
     @staticmethod
-    def primspec(i, sizes):
-        return torch.toffee.op("Reshape", i, shape=sizes, _outputs=(0, -1))
+    def primspec(g, i, sizes):
+        n = g.appendNode(g.create("Reshape", [i]).is_("shape", sizes))
+        real_out = g.appendNode(g.createSelect(n, 0))
+        g.appendNode(g.createSelect(n, 1))
+        return real_out
```
Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Do we have to return anything? We're mutating g anyway, right?

Comment (apaszke) [inline torch/autograd/_functions/tensor.py]: Can you leave a comment in here? 😕 This should be fixed in Toffee

File: torch/csrc/jit/interned_strings.h [hunk 4]
```diff
@@ -27,7 +27,20 @@ _(NumChunks) \
...
+_(is_test) \
+_(epsilon) \
+_(order) \
+_(momentum) \
+_(InPlaceOutputs) \
+_(kernels) \
+_(strides) \
+_(pads) \
+_(dilations) \
+_(group)
```
Comment (apaszke) [inline torch/csrc/jit/interned_strings.h:43]: I'm starting to worry that we're going to end up with name clashes because the `k` prefix is quite common. How about changing to to a `S` suffix?

File: torch/csrc/jit/ir.h [hunk 5]
```diff
@@ -233,6 +237,7 @@ struct Node : public Attributes<Node> {
   Graph* graph_;
   size_t unique_ = 0;          // unique id
   size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
+  std::string debug_name_;
```
Comment (apaszke) [inline torch/csrc/jit/ir.h:240]: Is there any reason why we don't want to keep that in attributes?

File: torch/csrc/jit/python_ir.cpp [hunk 6]
```diff
@@ -1,7 +1,10 @@
+#undef NDEBUG
 #include <pybind11/pybind11.h>
 // DO NOT REMOVE, this enables std containers to be recognized
 // with pybind11, removing the include disables the support
 #include <pybind11/stl.h>
+#define NDEBUG
```
Comment (apaszke) [inline torch/csrc/jit/python_ir.cpp]: ?

File: torch/csrc/toffee/export.cpp [hunk 7]
```diff
@@ -9,327 +9,259 @@
...
 std::string node_name(Node* n) {
-  return std::to_string(n->unique());
+  return n->uniqueName();
 }
 
-// Exports a graph to ToffeeIR
-std::string ExportGraph(std::shared_ptr<Graph>& g, const std::vector<at::Tensor> & initializers) {
-  toffee::GraphProto p_g;
+template<typename R, typename T>
+static std::vector<R> mapv(const std::vector<T> & inputs, std::function<R(const T &)> fn) {
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Awesome!! Would be nice to be able to use `auto` args in lambdas (C++14) 😕  Can you move this to `torch/utils/functional.h`? I'd also call it `fmap` and maybe template over some other containers later (might be useful)

File: torch/nn/_functions/thnn/auto.py [hunk 8]
```diff
@@ -29,7 +29,9 @@ def _make_function_class_criterion(class_name, update_output, update_grad_input,
 
     @staticmethod
     def primspec(*args, **kwargs):
-        return primspec_fn(*args, **kwargs)
+        a = primspec_fn(*args, **kwargs)
+        print("RETURNING " + repr(a))
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto.py]: Remove this

File: torch/csrc/toffee/export.cpp [hunk 9]
```diff
@@ -9,327 +9,259 @@
...
-            JIT_ASSERT(u.user->uses().empty());
-            continue;
-          }
-          p_n->add_output(node_name(u.user));
-        }
+  std::unordered_map<Node*, Node*> env;
+  ctx.graph = std::make_shared<Graph>();
+  for (auto input : g->inputs())
+    env[input] = ctx.graph->addInput()->setType(input->typeOption());
+  auto envFn = [&](Node * n) {
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: nit: I'd capture only the `env`

Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: nit: I'd just capture the `env`

File: torch/csrc/toffee/export.cpp [hunk 10]
```diff
@@ -9,327 +9,259 @@
...
+    return env.at(n);
+  };
+  // put the new outputs in our environment map, and
+  // copy the type from the input graph if they were not set by the
+  // primspec
+  auto setOutputs = [&](Node * node, const std::vector<Node*> & outputs) {
+    auto old_outputs = node->outputs();
+    JIT_ASSERTM(outputs.size() <= old_outputs.size(), "primspec produced too many outputs");
+    size_t i = 0;
+    for(auto & old : old_outputs) {
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp:49]: Maybe just change it to a regular `for` loop if you need a counter?

File: torch/csrc/toffee/export.cpp [hunk 11]
```diff
@@ -9,327 +9,259 @@
...
+      if(i >= outputs.size()) {
+        // primspecs do not deal with Handles at the moment
+        // so we map handles to Unused nodes
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp:52]: Ideally this would just raise an error. We'll only be able to convert ops that have traceable backward (unless no backward exists), and these don't have handles

File: torch/csrc/toffee/export.cpp [hunk 12]
```diff
@@ -9,327 +9,259 @@
...
+        auto outputs = fn->primspec(&ctx, mapv<Node*,Node*>(node->inputs(),envFn));
+        setOutputs(node,outputs);
       } else {
         throw std::runtime_error("CppOp doesn't define primspec " + value->name());
       }
     IR_ELSEIFM(PythonOp)
-      // NB: All inplace designations are dropped
+      auto pyobj = py::handle(value->pyobj.get());
+      if(!py::hasattr(pyobj, "primspec"))
+        throw std::runtime_error("PythonOp doesn't define primspec " + value->name());
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp:80]: Sooo we don't plan to replace init pass with this anytime soon? We can't raise errors for no primspecs then

File: torch/csrc/toffee/export.cpp [hunk 13]
```diff
@@ -9,327 +9,259 @@
...
 
-      toffee::NodeProto* p_n = generic_node();
+      py::object primspec_fn = pyobj.attr("primspec");
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp:82]: Is `py::object` a smart pointer?

File: torch/csrc/toffee/export.cpp [hunk 14]
```diff
@@ -9,327 +9,259 @@
...
+          JIT_ASSERT(back == n_i);
         } else {
           throw std::runtime_error("unexpected calling convention");
         }
-        PyTuple_SET_ITEM(py_primspec_args.get(), input_nr++, obj);
+        py_primspec_args[input_nr++] = obj;
       }
-      THPObjectPtr raw_output(PyObject_CallObject(primspec_fn, py_primspec_args));
-      if (!raw_output) {
+      py::object raw_output = py::reinterpret_borrow<py::object>(PyObject_CallObject(primspec_fn.ptr(), py_primspec_args.ptr()));
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Not sure what it does, but `CallObject` returns a new reference

File: torch/csrc/toffee/export.cpp [hunk 15]
```diff
@@ -9,327 +9,259 @@
...
+      setOutputs(node,outputs);
+    IR_ELSE()
+      if(node->kind() == kConstant && node->t(kValue).defined()) {
+        throw std::runtime_error("Constant not supported yet");
+      }
+      auto n_ = ctx.graph->createClone(node, envFn);
+      if(node->hasMultipleOutputs()) {
+        int i = 0;
+        for(auto s : node->uses()) {
+          env[s.user] = ctx.graph->createSelect(n_,i++);
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Don't you want to copy types here?

**Review by apaszke — COMMENTED** [review]
Mostly looks good, but I have a few questions before merging

File: torch/autograd/_functions/tensor.py (hunk 3 above)
Comment (zdevito) [reply torch/autograd/_functions/tensor.py]: You return as many outputs as the original thing had. This allows you to replace a single node with an entire subgraph if you (e.g. doing a transpose, then a conv, and returning the results of the conv)

Comment (zdevito) [reply torch/autograd/_functions/tensor.py]: yes

File: torch/csrc/jit/ir.h (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/ir.h:240]: The attributes are suppose to stay with the toffeeIR schema, so I didn't want people to get their debug names leaking into Toffee. I think Toffee should have debug names for nodes as well (do they have then already?).

File: torch/csrc/toffee/export.cpp (hunk 7 above)
Comment (zdevito) [reply torch/csrc/toffee/export.cpp]: done

File: torch/csrc/toffee/export.cpp (hunk 13 above)
Comment (zdevito) [reply torch/csrc/toffee/export.cpp:82]: py::object is a shared_ptr. py::handle is the raw PyObject with no reference counts. You convert with py::reinterpret_borrow or py::reinterpret_steal

File: torch/csrc/toffee/export.cpp (hunk 11 above)
Comment (zdevito) [reply torch/csrc/toffee/export.cpp:52]: yes, future TODO for when we get this merged with init_pass. For now these objects do have Handles attached and we need to ignore them

File: torch/csrc/toffee/export.cpp (hunk 12 above)
Comment (zdevito) [reply torch/csrc/toffee/export.cpp:80]: We do want to replace it, but not in this commit. There are a few places like this that will need to be sorted out.

**Review by apaszke — APPROVED** [review]
